solveSimpleWanteds

```
    solveSimpleWanteds: too many iterations (limit = 4)
      Set limit with -fconstraint-solver-iterations=n; n=0 for no limit
      Simples = {[WD] $dMonad_a1CfU {0}:: Monad m0 (CNonCanonical),
                 [WD] $d(%,%)_a1Cg0 {0}:: Member (Lift IO) r0 (CNonCanonical),
                 [WD] hole{co_a1Cg4} {0}:: r0
                                           ~ '[Input WK2.WebView, RPC, Lift m0] (CNonCanonical)}
      WC = WC {wc_simple =
                 [WD] hole{co_a1Cwc} {1}:: Lift IO ~ Lift m0 (CNonCanonical)
                 [WD] hole{co_a1Cwa} {1}:: Lift IO ~ Lift m0 (CNonCanonical)
                 [W] $dMonad_a1CfU {0}:: Monad m0 (CDictCan(psc))
                 [W] $dFind_a1Cvr {1}:: Polysemy.Internal.Union.Find
                                          '[Lift m0] (Lift IO) (CDictCan)
                 [WD] hole{co_a1Cw8} {7}:: Polysemy.Internal.Union.IndexOf
                                             '[Lift m0]
                                             (Polysemy.Internal.Union.Found '[Lift m0] (Lift IO))
                                           ~ Lift IO (CNonCanonical)
                 [W] hole{co_a1Cg4} {0}:: r0
                                          ~ '[Input WK2.WebView, RPC, Lift m0] (CNonCanonical)
                 [WD] hole{co_a1Cw1} {2}:: m0 ~ IO (CNonCanonical)
                 [D] _ {0}:: r0 ~ '[Input WK2.WebView, RPC, Lift IO] (CNonCanonical)
                 [D] _ {6}:: Polysemy.Internal.Union.Found '[Lift m0] (Lift IO)
                             ~ 'Polysemy.Internal.Union.Z (CNonCanonical)}
```

Also in `ExampleSpec` when using `runResource` instead of `runResourceInIO`:

```
/home/sandy/prj/polysemy/polysemy-plugin/test/ExampleSpec.hs:1:1: error:
    solveSimpleWanteds: too many iterations (limit = 4)
      Set limit with -fconstraint-solver-iterations=n; n=0 for no limit
      Simples = {[WD] irred_ag9W {1}:: Polysemy.Internal.CustomErrors.Break
                                         (TypeError ...)
                                         (Polysemy.Internal.Union.IndexOf
                                            r0
                                            (Polysemy.Internal.Union.Found
                                               r0 (Lift IO))) (CIrredCan(sol)),
                 [WD] $dFind_ag9U {1}:: Polysemy.Internal.Union.Find
                                          r0 (Lift IO) (CDictCan(psc)),
                 [WD] hole{co_ag9Z} {4}:: Polysemy.Internal.Union.IndexOf
                                            r0 (Polysemy.Internal.Union.Found r0 (Lift IO))
                                          ~ Lift IO (CNonCanonical)}
      WC = WC {wc_simple =           
                 [WD] hole{co_agbC} {1}:: Lift IO ~ e0 (CNonCanonical)
                 [WD] hole{co_agbA} {1}:: Lift IO ~ e0 (CNonCanonical)
                 [W] irred_agby {1}:: Polysemy.Internal.CustomErrors.Break
                                        (TypeError ...)
                                        (Polysemy.Internal.Union.IndexOf
                                           '[e0, Lift IO]
                                           (Polysemy.Internal.Union.Found
                                              '[e0, Lift IO] (Lift IO))) (CIrredCan(sol))
                 [W] $dFind_ag9U {1}:: Polysemy.Internal.Union.Find
                                         '[e0, Lift IO] (Lift IO) (CDictCan(psc))
                 [WD] hole{co_agbr} {8}:: Polysemy.Internal.Union.IndexOf
                                            '[e0, Lift IO]
                                            (Polysemy.Internal.Union.Found '[e0, Lift IO] (Lift IO))
                                          ~ Lift IO (CNonCanonical)
                 [WD] hole{co_agbl} {2}:: e0 ~ Lift IO (CNonCanonical)
                 [D] _ {1}:: (TypeError ...) ~ (TypeError ...) (CNonCanonical)
                 [D] _ {7}:: Polysemy.Internal.Union.Found '[e0, Lift IO] (Lift IO)
                             ~ 'Polysemy.Internal.Union.Z (CNonCanonical)
                 [D] _ {1}:: Polysemy.Internal.CustomErrors.Break
                               (TypeError ...)
                               (Polysemy.Internal.Union.IndexOf
                                  '[e0, Lift IO]
                                  (Polysemy.Internal.Union.Found '[e0, Lift IO] (Lift IO)))
                             ~ () :: Constraint (CNonCanonical)}
```

I'm not sure if this is related, but I'm getting such an error with this snippet (heavily minimized, so minimized it wouldn't even compile anymore):
```haskell
{-# LANGUAGE DataKinds  #-}
{-# LANGUAGE RankNTypes #-}

module Arvy.Algorithm where

import           Polysemy
import           Polysemy.State

foo :: Sem '[State i] ()
foo = put ()
```

```
lib/Arvy/Algorithm.hs:1:1: error:
    solveSimpleWanteds: too many iterations (limit = 4)
      Set limit with -fconstraint-solver-iterations=n; n=0 for no limit
      Simples = {[WD] $d(%,,%)_a4BZ {0}:: Member
                                            (State s0) r0 (CNonCanonical)}
      WC = WC {wc_simple =
                 [WD] hole{co_a50e} {1}:: State () ~ State i (CNonCanonical)
                 [W] irred_a4ZQ {1}:: Polysemy.Internal.CustomErrors.Break
                                        (TypeError ...)
                                        (Polysemy.Internal.Union.IndexOf
                                           '[State i]
                                           (Polysemy.Internal.Union.Found
                                              '[State i] (State ()))) (CIrredCan(sol))
                 [W] $dFind_a4Ki {1}:: Polysemy.Internal.Union.Find
                                         '[State i] (State ()) (CDictCan)
                 [WD] hole{co_a4YG} {7}:: Polysemy.Internal.Union.IndexOf
                                            '[State i]
                                            (Polysemy.Internal.Union.Found '[State i] (State ()))
                                          ~ State () (CNonCanonical)
                 [WD] hole{co_a4Ym} {2}:: i ~ () (CNonCanonical)
                 [D] _ {1}:: (TypeError ...) ~ (TypeError ...) (CNonCanonical)
                 [D] _ {6}:: Polysemy.Internal.Union.Found '[State i] (State ())
                             ~ 'Polysemy.Internal.Union.Z (CNonCanonical)
                 [D] _ {1}:: Polysemy.Internal.CustomErrors.Break
                               (TypeError ...)
                               (Polysemy.Internal.Union.IndexOf
                                  '[State i] (Polysemy.Internal.Union.Found '[State i] (State ())))
                             ~ () :: Constraint (CNonCanonical)}
```

And here's a snippet which *should* compile, which is more like my original code:
```haskell
{-# LANGUAGE GADTs #-}
{-# LANGUAGE DataKinds  #-}
{-# LANGUAGE RankNTypes #-}

module Arvy.Algorithm where

import           Polysemy
import           Polysemy.State

data Bar = forall s . Bar (forall i . Sem '[State s] ())

bar :: Bar
bar = Bar (put ())
```

```
lib/Arvy/Algorithm.hs:1:1: error:
    solveSimpleWanteds: too many iterations (limit = 4)
      Set limit with -fconstraint-solver-iterations=n; n=0 for no limit
      Simples = {[WD] $d(%,,%)_a4Cq {0}:: Member
                                            (State s0) r0 (CNonCanonical)}
      WC = WC {wc_simple =
                 [WD] hole{co_a52Y} {1}:: State () ~ State s0 (CNonCanonical)
                 [W] irred_a51x {1}:: Polysemy.Internal.CustomErrors.Break
                                        (TypeError ...)
                                        (Polysemy.Internal.Union.IndexOf
                                           '[State s0]
                                           (Polysemy.Internal.Union.Found
                                              '[State s0] (State ()))) (CIrredCan(sol))
                 [W] $dFind_a4KJ {1}:: Polysemy.Internal.Union.Find
                                         '[State s0] (State ()) (CDictCan)
                 [WD] hole{co_a50w} {7}:: Polysemy.Internal.Union.IndexOf
                                            '[State s0]
                                            (Polysemy.Internal.Union.Found '[State s0] (State ()))
                                          ~ State () (CNonCanonical)
                 [WD] hole{co_a50r} {2}:: s0 ~ () (CNonCanonical)
                 [D] _ {1}:: (TypeError ...) ~ (TypeError ...) (CNonCanonical)
                 [D] _ {6}:: Polysemy.Internal.Union.Found '[State s0] (State ())
                             ~ 'Polysemy.Internal.Union.Z (CNonCanonical)
                 [D] _ {1}:: Polysemy.Internal.CustomErrors.Break
                               (TypeError ...)
                               (Polysemy.Internal.Union.IndexOf
                                  '[State s0]
                                  (Polysemy.Internal.Union.Found '[State s0] (State ())))
                             ~ () :: Constraint (CNonCanonical)}
```

Today when I tried to update my local polysemy version, I encountered 2 different polysemy-plugin errors. One test failure when updating to 0.3.0.0, and in addition a `solveSimpleWanteds` failure when updating to 0.4.0.0.

- The former happens with polysemy-plugin 0.2.0.2 and polysemy 0.3.0.0, [full log](https://paste.infinisil.com/qELqikEXls.log), relevant part:

	<details>

	```
	Failures:

		test/InlineRecursiveCallsSpec.hs:26:25:
		1) InlineRecursiveCalls, inlining recursive calls, should explicitly break recursion
				 uncaught exception: ErrorCall
				 test/InlineRecursiveCallsSpec.hs:21:21: recursive === mutual failed:
						 LHS:
								 recursive
									 :: forall (e :: (* -> *) -> * -> *) s (r :: [(* -> *)
																																-> * -> *]) a.
											(forall x (m :: * -> *). e m x -> StateT s (Sem r) x)
											-> s
											-> Sem (e : r) a
											-> forall (m :: * -> *).
												 Monad m =>
												 (forall x. Union r (Sem r) x -> m x) -> m (s, a)
								 recursive
									 = \ (@ (e :: (* -> *) -> * -> *))
											 (@ s)
											 (@ (r :: [(* -> *) -> * -> *]))
											 (@ a)
											 (f :: forall x (m :: * -> *). e m x -> StateT s (Sem r) x)
											 (s :: s)
											 (ds :: Sem (e : r) a)
											 (@ (m :: * -> *))
											 ($dMonad :: Monad m) ->
											 let {
												 $dMonad :: Monad (StateT s m)
												 $dMonad = $fMonadStateT @ m @ s $dMonad } in
											 let {
												 $dApplicative :: Applicative m
												 $dApplicative = $p1Monad @ m $dMonad } in
											 let {
												 $dFunctor :: Functor m
												 $dFunctor = $p1Applicative @ m $dApplicative } in
											 \ (k :: forall x. Union r (Sem r) x -> m x) ->
												 fmap
													 @ m
													 $dFunctor
													 @ (a, s)
													 @ (s, a)
													 (swap @ a @ s)
													 ((((ds `cast` <Co:13>)
																@ (StateT s m)
																$dMonad
																(\ (@ x) (u :: Union (e : r) (Sem (e : r)) x) ->
																	 case u of { Union @ n p a1 ->
																	 case p of {
																		 SZ co ->
																			 case a1 `cast` <Co:48> of
																			 { Yo @ f @ m1 @ a1 $dFunctor e z ds y ds ->
																			 let {
																				 lvl :: (a1, s) -> (x, s)
																				 lvl
																					 = \ (w :: (a1, s)) ->
																							 (y (<$
																										 @ f
																										 $dFunctor
																										 @ a1
																										 @ ()
																										 (case w of { (a1, s') -> a1 })
																										 z),
																								case w of { (a1, s') -> s' }) } in
																			 let {
																				 m1 :: StateT s (Sem r) a1
																				 m1 = f @ a1 @ m1 e } in
																			 (\ (s1 :: s) ->
																					fmap
																						@ m
																						$dFunctor
																						@ (a1, s)
																						@ (x, s)
																						lvl
																						((((m1 `cast` <Co:5>) s1) `cast` <Co:5>) @ m $dMonad k))
																			 `cast` <Co:5>
																			 };
																		 SS @ n1 co n2 ->
																			 case a1 `cast` <Co:49> of
																			 { Yo @ f1 @ m1 @ a1 $dFunctor3 e1 s2 nt f2 v1 ->
																			 (\ (s' :: s) ->
																					k @ (x, s)
																						(Union
																							 @ r
																							 @ (Sem r)
																							 @ (x, s)
																							 @ n1
																							 n2
																							 (Yo
																									@ (IndexOf r n1)
																									@ (Sem r)
																									@ (x, s)
																									@ (Compose ((,) s) f1)
																									@ m1
																									@ a1
																									($fFunctorCompose
																										 @ ((,) s) @ f1 ($fFunctor(,) @ s) $dFunctor3)
																									e1
																									((s', s2) `cast` <Co:10>)
																									(\ (@ x) (x1 :: Compose ((,) s) f1 (m1 x)) ->
																										 let {
																											 p :: (s, Sem (e : r) (f1 x))
																											 p = case x1 `cast` <Co:10> of { (x, y) ->
																													 (x, nt @ x y)
																													 } } in
																										 let {
																											 ds :: s
																											 ds = case p of { (x, ds) -> x } } in
																										 let {
																											 ds :: Sem (e : r) (f1 x)
																											 ds = case p of { (ds, y) -> y } } in
																										 (\ (@ (m :: * -> *))
																												($dMonad :: Monad m)
																												(eta :: forall x. Union r (Sem r) x -> m x) ->
																												fmap
																													@ m
																													($p1Applicative @ m ($p1Monad @ m $dMonad))
																													@ (s, f1 x)
																													@ (Compose ((,) s) f1 x)
																													((poly_f4 @ x @ f1 @ s) `cast` <Co:15>)
																													(((recursive @ e @ s @ r @ (f1 x) f ds ds)
																														`cast` <Co:6>)
																														 @ m $dMonad eta))
																										 `cast` <Co:12>)
																									(\ (x :: Compose ((,) s) f1 a1) ->
																										 case x `cast` <Co:9> of { (x, y) -> (f2 y, x) })
																									(\ (@ x) (eta :: Compose ((,) s) f1 x) ->
																										 v1 @ x (snd @ s @ (f1 x) (eta `cast` <Co:9>))))))
																			 `cast` <Co:5>
																			 }
																	 }
																	 }))
														 `cast` <Co:4>)
															s)

								 recursive
									 :: forall (e :: (* -> *) -> * -> *) s (r :: [(* -> *)
																																-> * -> *]) a.
											(forall x (m :: * -> *). e m x -> StateT s (Sem r) x)
											-> s -> Sem (e : r) a -> Sem r (s, a)
								 recursive = recursive `cast` <Co:62>

								 recursive
									 :: forall (e :: (* -> *) -> * -> *) s (r :: [(* -> *)
																																-> * -> *]) a.
											(forall x (m :: * -> *). e m x -> StateT s (Sem r) x)
											-> s -> Sem (e : r) a -> Sem r (s, a)
								 recursive = recursive

						 RHS:
								 mutual
									 :: forall (e :: (* -> *) -> * -> *) s (r :: [(* -> *)
																																-> * -> *]) a.
											(forall x (m :: * -> *). e m x -> StateT s (Sem r) x)
											-> s -> Sem (e : r) a -> Sem r (s, a)
								 mutual = mutual `cast` <Co:62>

								 mutual2
									 :: forall (e :: (* -> *) -> * -> *) s (r :: [(* -> *)
																																-> * -> *]) a.
											(forall x (m :: * -> *). e m x -> StateT s (Sem r) x)
											-> s -> Sem (e : r) a -> Sem r (s, a)
								 mutual2 = mutual

								 mutual
									 :: forall (e :: (* -> *) -> * -> *) s (r :: [(* -> *)
																																-> * -> *]) a.
											(forall x (m :: * -> *). e m x -> StateT s (Sem r) x)
											-> s
											-> Sem (e : r) a
											-> forall (m :: * -> *).
												 Monad m =>
												 (forall x. Union r (Sem r) x -> m x) -> m (s, a)
								 mutual
									 = \ (@ (e :: (* -> *) -> * -> *))
											 (@ s)
											 (@ (r :: [(* -> *) -> * -> *]))
											 (@ a)
											 (f :: forall x (m :: * -> *). e m x -> StateT s (Sem r) x)
											 (s :: s)
											 (ds :: Sem (e : r) a)
											 (@ (m :: * -> *))
											 ($dMonad :: Monad m) ->
											 let {
												 $dMonad :: Monad (StateT s m)
												 $dMonad = $fMonadStateT @ m @ s $dMonad } in
											 let {
												 $dApplicative :: Applicative m
												 $dApplicative = $p1Monad @ m $dMonad } in
											 let {
												 $dFunctor :: Functor m
												 $dFunctor = $p1Applicative @ m $dApplicative } in
											 \ (k :: forall x. Union r (Sem r) x -> m x) ->
												 fmap
													 @ m
													 $dFunctor
													 @ (a, s)
													 @ (s, a)
													 (swap @ a @ s)
													 ((((ds `cast` <Co:13>)
																@ (StateT s m)
																$dMonad
																(\ (@ x) (u :: Union (e : r) (Sem (e : r)) x) ->
																	 case u of { Union @ n p a1 ->
																	 case p of {
																		 SZ co ->
																			 case a1 `cast` <Co:48> of
																			 { Yo @ f @ m1 @ a1 $dFunctor e z ds y ds ->
																			 let {
																				 lvl :: (a1, s) -> (x, s)
																				 lvl
																					 = \ (w :: (a1, s)) ->
																							 (y (<$
																										 @ f
																										 $dFunctor
																										 @ a1
																										 @ ()
																										 (case w of { (a1, s') -> a1 })
																										 z),
																								case w of { (a1, s') -> s' }) } in
																			 let {
																				 m1 :: StateT s (Sem r) a1
																				 m1 = f @ a1 @ m1 e } in
																			 (\ (s1 :: s) ->
																					fmap
																						@ m
																						$dFunctor
																						@ (a1, s)
																						@ (x, s)
																						lvl
																						((((m1 `cast` <Co:5>) s1) `cast` <Co:5>) @ m $dMonad k))
																			 `cast` <Co:5>
																			 };
																		 SS @ n1 co n2 ->
																			 case a1 `cast` <Co:49> of
																			 { Yo @ f1 @ m1 @ a1 $dFunctor3 e1 s2 nt f2 v1 ->
																			 (\ (s' :: s) ->
																					k @ (x, s)
																						(Union
																							 @ r
																							 @ (Sem r)
																							 @ (x, s)
																							 @ n1
																							 n2
																							 (Yo
																									@ (IndexOf r n1)
																									@ (Sem r)
																									@ (x, s)
																									@ (Compose ((,) s) f1)
																									@ m1
																									@ a1
																									($fFunctorCompose
																										 @ ((,) s) @ f1 ($fFunctor(,) @ s) $dFunctor3)
																									e1
																									((s', s2) `cast` <Co:10>)
																									(\ (@ x) (x1 :: Compose ((,) s) f1 (m1 x)) ->
																										 let {
																											 p :: (s, Sem (e : r) (f1 x))
																											 p = case x1 `cast` <Co:10> of { (x, y) ->
																													 (x, nt @ x y)
																													 } } in
																										 let {
																											 ds :: s
																											 ds = case p of { (x, ds) -> x } } in
																										 let {
																											 ds :: Sem (e : r) (f1 x)
																											 ds = case p of { (ds, y) -> y } } in
																										 (\ (@ (m :: * -> *))
																												($dMonad :: Monad m)
																												(eta :: forall x. Union r (Sem r) x -> m x) ->
																												fmap
																													@ m
																													($p1Applicative @ m ($p1Monad @ m $dMonad))
																													@ (s, f1 x)
																													@ (Compose ((,) s) f1 x)
																													((poly_f4 @ x @ f1 @ s) `cast` <Co:15>)
																													(((mutual2 @ e @ s @ r @ (f1 x) f ds ds)
																														`cast` <Co:6>)
																														 @ m $dMonad eta))
																										 `cast` <Co:12>)
																									(\ (x :: Compose ((,) s) f1 a1) ->
																										 case x `cast` <Co:9> of { (x, y) -> (f2 y, x) })
																									(\ (@ x) (eta :: Compose ((,) s) f1 x) ->
																										 v1 @ x (snd @ s @ (f1 x) (eta `cast` <Co:9>))))))
																			 `cast` <Co:5>
																			 }
																	 }
																	 }))
														 `cast` <Co:4>)
															s)

				 CallStack (from HasCallStack):
					 error, called at test/InlineRecursiveCallsSpec.hs:26:25 in main:InlineRecursiveCallsSpec

		To rerun use: --match "/InlineRecursiveCalls/inlining recursive calls/should explicitly break recursion/"

	Randomized with seed 775772120

	Finished in 0.0104 seconds
	27 examples, 1 failure
	Test suite polysemy-plugin-test: FAIL
	Test suite logged to:
	dist/test/polysemy-plugin-0.2.0.2-polysemy-plugin-test.log
	0 of 1 test suites (0 of 1 test cases) passed.
	```
	</details>

    This is probably not related to this issue, but who knows.


- The latter (and the one relevant to this issue), happens with polysemy-plugin 0.2.0.2 and polysemy 0.4.0.0, [full log](https://paste.infinisil.com/WEBU2OSQm0.log), relevant part near beginning:

  ```
  test/ExampleSpec.hs:1:1: error:
      solveSimpleWanteds: too many iterations (limit = 4)
        Set limit with -fconstraint-solver-iterations=n; n=0 for no limit
        Simples = {[WD] irred_asTR {1}:: Polysemy.Internal.CustomErrors.Break
                                           (TypeError ...)
                                           (Polysemy.Internal.Union.IndexOf
                                              r0
                                              (Polysemy.Internal.Union.Found
                                                 r0 (Lift IO))) (CIrredCan(sol)),
                   [WD] $dFind_asTy {1}:: Polysemy.Internal.Union.Find
                                            r0 (Lift IO) (CDictCan(psc)),
                   [WD] hole{co_asVA} {4}:: Polysemy.Internal.Union.IndexOf
                                              r0 (Polysemy.Internal.Union.Found r0 (Lift IO))
                                            ~ Lift IO (CNonCanonical)}
        WC = WC {wc_simple =
                   [WD] hole{co_atGE} {1}:: Lift IO ~ e0 (CNonCanonical)
                   [WD] hole{co_atGy} {1}:: Lift IO ~ e0 (CNonCanonical)
                   [W] irred_atEh {1}:: Polysemy.Internal.CustomErrors.Break
                                          (TypeError ...)
                                          (Polysemy.Internal.Union.IndexOf
                                             '[e0, Lift IO]
                                             (Polysemy.Internal.Union.Found
                                                '[e0, Lift IO] (Lift IO))) (CIrredCan(sol))
                   [W] $dFind_asTy {1}:: Polysemy.Internal.Union.Find
                                           '[e0, Lift IO] (Lift IO) (CDictCan(psc))
                   [WD] hole{co_atBP} {8}:: Polysemy.Internal.Union.IndexOf
                                              '[e0, Lift IO]
                                              (Polysemy.Internal.Union.Found '[e0, Lift IO] (Lift IO))
                                            ~ Lift IO (CNonCanonical)
                   [WD] hole{co_atBi} {2}:: e0 ~ Lift IO (CNonCanonical)
                   [D] _ {1}:: (TypeError ...) ~ (TypeError ...) (CNonCanonical)
                   [D] _ {7}:: Polysemy.Internal.Union.Found '[e0, Lift IO] (Lift IO)
                               ~ 'Polysemy.Internal.Union.Z (CNonCanonical)
                   [D] _ {1}:: Polysemy.Internal.CustomErrors.Break
                                 (TypeError ...)
                                 (Polysemy.Internal.Union.IndexOf
                                    '[e0, Lift IO]
                                    (Polysemy.Internal.Union.Found '[e0, Lift IO] (Lift IO)))
                               ~ () :: Constraint (CNonCanonical)}
    |
  1 | {-# LANGUAGE TemplateHaskell #-}
    | ^
  ```

  And inbetween, the same test failure from before is visible, but it's not neatly printed because it exits early due to the other error:

  <details>

  ```
  ==================== Tidy Core ====================
  Result size of Tidy Core
    = {terms: 447, types: 1,819, coercions: 619, joins: 0/16}

  -- RHS size: {terms: 3, types: 0, coercions: 0, joins: 0/0}
  inspect_21_21_6989586621679061611 [InlPrag=NOINLINE] :: Result
  [GblId,
   Unf=Unf{Src=InlineStable, TopLvl=True, Value=False, ConLike=False,
           WorkFree=True, Expandable=True,
           Guidance=ALWAYS_IF(arity=0,unsat_ok=True,boring_ok=True)
           Tmpl= Test.Inspection.didNotRunPluginError}]
  inspect_21_21_6989586621679061611
    = Test.Inspection.Failure
        (ghc-prim-0.5.3:GHC.CString.unpackCString#
           "test/InlineRecursiveCallsSpec.hs:21:21: recursive === mutual failed:\n\
           \    LHS:\n\
           \        recursive\n\
           \          :: forall (e :: (* -> *) -> * -> *) s (r :: [(* -> *)\n\
           \                                                       -> * -> *]) a.\n\
           \             (forall x (m :: * -> *). e m x -> StateT s (Sem r) x)\n\
           \             -> s\n\
           \             -> Sem (e : r) a\n\
           \             -> forall (m :: * -> *).\n\
           \                Monad m =>\n\
           \                (forall x. Union r (Sem r) x -> m x) -> m (s, a)\n\
           \        recursive\n\
           \          = \\ (@ (e :: (* -> *) -> * -> *))\n\
           \              (@ s)\n\
           \              (@ (r :: [(* -> *) -> * -> *]))\n\
           \              (@ a)\n\
           \              (f :: forall x (m :: * -> *). e m x -> StateT s (Sem r) x)\n\
           \              (s :: s)\n\
           \              (ds :: Sem (e : r) a)\n\
           \              (@ (m :: * -> *))\n\
           \              ($dMonad :: Monad m) ->\n\
           \              let {\n\
           \                $dMonad :: Monad (StateT s m)\n\
           \                $dMonad = $fMonadStateT @ m @ s $dMonad } in\n\
           \              let {\n\
           \                $dApplicative :: Applicative m\n\
           \                $dApplicative = $p1Monad @ m $dMonad } in\n\
           \              let {\n\
           \                $dFunctor :: Functor m\n\
           \                $dFunctor = $p1Applicative @ m $dApplicative } in\n\
           \              \\ (k :: forall x. Union r (Sem r) x -> m x) ->\n\
           \                fmap\n\
           \                  @ m\n\
           \                  $dFunctor\n\
           \                  @ (a, s)\n\
           \                  @ (s, a)\n\
           \                  (swap @ a @ s)\n\
           \                  ((((ds `cast` <Co:13>)\n\
           \                       @ (StateT s m)\n\
           \                       $dMonad\n\
           \                       (\\ (@ x) (u :: Union (e : r) (Sem (e : r)) x) ->\n\
           \                          case u of { Union @ n p a1 ->\n\
           \                          case p of {\n\
           \                            SZ co ->\n\
           \                              case a1 `cast` <Co:48> of\n\
           \                              { Yo @ f @ m1 @ a1 $dFunctor e z ds y ds ->\n\
           \                              let {\n\
           \                                lvl :: (a1, s) -> (x, s)\n\
           \                                lvl\n\
           \                                  = \\ (w :: (a1, s)) ->\n\
           \                                      (y (<$\n\
           \                                            @ f\n\
           \                                            $dFunctor\n\
           \                                            @ a1\n\
           \                                            @ ()\n\
           \                                            (case w of { (a1, s') -> a1 })\n\
           \                                            z),\n\
           \                                       case w of { (a1, s') -> s' }) } in\n\
           \                              let {\n\
           \                                m1 :: StateT s (Sem r) a1\n\
           \                                m1 = f @ a1 @ m1 e } in\n\
           \                              (\\ (s1 :: s) ->\n\
           \                                 fmap\n\
           \                                   @ m\n\
           \                                   $dFunctor\n\
           \                                   @ (a1, s)\n\
           \                                   @ (x, s)\n\
           \                                   lvl\n\
           \                                   ((((m1 `cast` <Co:5>) s1) `cast` <Co:5>) @ m $dMonad k))\n\
           \                              `cast` <Co:5>\n\
           \                              };\n\
           \                            SS @ n1 co n2 ->\n\
           \                              case a1 `cast` <Co:49> of\n\
           \                              { Yo @ f1 @ m1 @ a1 $dFunctor3 e1 s2 nt f2 v1 ->\n\
           \                              (\\ (s' :: s) ->\n\
           \                                 k @ (x, s)\n\
           \                                   (Union\n\
           \                                      @ r\n\
           \                                      @ (Sem r)\n\
           \                                      @ (x, s)\n\
           \                                      @ n1\n\
           \                                      n2\n\
           \                                      (Yo\n\
           \                                         @ (IndexOf r n1)\n\
           \                                         @ (Sem r)\n\
           \                                         @ (x, s)\n\
           \                                         @ (Compose ((,) s) f1)\n\
           \                                         @ m1\n\
           \                                         @ a1\n\
           \                                         ($fFunctorCompose\n\
           \                                            @ ((,) s) @ f1 ($fFunctor(,) @ s) $dFunctor3)\n\
           \                                         e1\n\
           \                                         ((s', s2) `cast` <Co:10>)\n\
           \                                         (\\ (@ x) (x1 :: Compose ((,) s) f1 (m1 x)) ->\n\
           \                                            let {\n\
           \                                              p :: (s, Sem (e : r) (f1 x))\n\
           \                                              p = case x1 `cast` <Co:10> of { (x, y) ->\n\
           \                                                  (x, nt @ x y)\n\
           \                                                  } } in\n\
           \                                            let {\n\
           \                                              ds :: s\n\
           \                                              ds = case p of { (x, ds) -> x } } in\n\
           \                                            let {\n\
           \                                              ds :: Sem (e : r) (f1 x)\n\
           \                                              ds = case p of { (ds, y) -> y } } in\n\
           \                                            (\\ (@ (m :: * -> *))\n\
           \                                               ($dMonad :: Monad m)\n\
           \                                               (eta :: forall x. Union r (Sem r) x -> m x) ->\n\
           \                                               fmap\n\
           \                                                 @ m\n\
           \                                                 ($p1Applicative @ m ($p1Monad @ m $dMonad))\n\
           \                                                 @ (s, f1 x)\n\
           \                                                 @ (Compose ((,) s) f1 x)\n\
           \                                                 ((poly_f4 @ x @ f1 @ s) `cast` <Co:15>)\n\
           \                                                 (((recursive @ e @ s @ r @ (f1 x) f ds ds)\n\
           \                                                   `cast` <Co:6>)\n\
           \                                                    @ m $dMonad eta))\n\
           \                                            `cast` <Co:12>)\n\
           \                                         (\\ (x :: Compose ((,) s) f1 a1) ->\n\
           \                                            case x `cast` <Co:9> of { (x, y) -> (f2 y, x) })\n\
           \                                         (\\ (@ x) (eta :: Compose ((,) s) f1 x) ->\n\
           \                                            v1 @ x (snd @ s @ (f1 x) (eta `cast` <Co:9>))))))\n\
           \                              `cast` <Co:5>\n\
           \                              }\n\
           \                          }\n\
           \                          }))\n\
           \                    `cast` <Co:4>)\n\
           \                     s)\n\
           \        \n\
           \        recursive\n\
           \          :: forall (e :: (* -> *) -> * -> *) s (r :: [(* -> *)\n\
           \                                                       -> * -> *]) a.\n\
           \             (forall x (m :: * -> *). e m x -> StateT s (Sem r) x)\n\
           \             -> s -> Sem (e : r) a -> Sem r (s, a)\n\
           \        recursive = recursive `cast` <Co:62>\n\
           \        \n\
           \        recursive\n\
           \          :: forall (e :: (* -> *) -> * -> *) s (r :: [(* -> *)\n\
           \                                                       -> * -> *]) a.\n\
           \             (forall x (m :: * -> *). e m x -> StateT s (Sem r) x)\n\
           \             -> s -> Sem (e : r) a -> Sem r (s, a)\n\
           \        recursive = recursive\n\
           \        \n\
           \    RHS:\n\
           \        mutual\n\
           \          :: forall (e :: (* -> *) -> * -> *) s (r :: [(* -> *)\n\
           \                                                       -> * -> *]) a.\n\
           \             (forall x (m :: * -> *). e m x -> StateT s (Sem r) x)\n\
           \             -> s -> Sem (e : r) a -> Sem r (s, a)\n\
           \        mutual = mutual `cast` <Co:62>\n\
           \        \n\
           \        mutual2\n\
           \          :: forall (e :: (* -> *) -> * -> *) s (r :: [(* -> *)\n\
           \                                                       -> * -> *]) a.\n\
           \             (forall x (m :: * -> *). e m x -> StateT s (Sem r) x)\n\
           \             -> s -> Sem (e : r) a -> Sem r (s, a)\n\
           \        mutual2 = mutual\n\
           \        \n\
           \        mutual\n\
           \          :: forall (e :: (* -> *) -> * -> *) s (r :: [(* -> *)\n\
           \                                                       -> * -> *]) a.\n\
           \             (forall x (m :: * -> *). e m x -> StateT s (Sem r) x)\n\
           \             -> s\n\
           \             -> Sem (e : r) a\n\
           \             -> forall (m :: * -> *).\n\
           \                Monad m =>\n\
           \                (forall x. Union r (Sem r) x -> m x) -> m (s, a)\n\
           \        mutual\n\
           \          = \\ (@ (e :: (* -> *) -> * -> *))\n\
           \              (@ s)\n\
           \              (@ (r :: [(* -> *) -> * -> *]))\n\
           \              (@ a)\n\
           \              (f :: forall x (m :: * -> *). e m x -> StateT s (Sem r) x)\n\
           \              (s :: s)\n\
           \              (ds :: Sem (e : r) a)\n\
           \              (@ (m :: * -> *))\n\
           \              ($dMonad :: Monad m) ->\n\
           \              let {\n\
           \                $dMonad :: Monad (StateT s m)\n\
           \                $dMonad = $fMonadStateT @ m @ s $dMonad } in\n\
           \              let {\n\
           \                $dApplicative :: Applicative m\n\
           \                $dApplicative = $p1Monad @ m $dMonad } in\n\
           \              let {\n\
           \                $dFunctor :: Functor m\n\
           \                $dFunctor = $p1Applicative @ m $dApplicative } in\n\
           \              \\ (k :: forall x. Union r (Sem r) x -> m x) ->\n\
           \                fmap\n\
           \                  @ m\n\
           \                  $dFunctor\n\
           \                  @ (a, s)\n\
           \                  @ (s, a)\n\
           \                  (swap @ a @ s)\n\
           \                  ((((ds `cast` <Co:13>)\n\
           \                       @ (StateT s m)\n\
           \                       $dMonad\n\
           \                       (\\ (@ x) (u :: Union (e : r) (Sem (e : r)) x) ->\n\
           \                          case u of { Union @ n p a1 ->\n\
           \                          case p of {\n\
           \                            SZ co ->\n\
           \                              case a1 `cast` <Co:48> of\n\
           \                              { Yo @ f @ m1 @ a1 $dFunctor e z ds y ds ->\n\
           \                              let {\n\
           \                                lvl :: (a1, s) -> (x, s)\n\
           \                                lvl\n\
           \                                  = \\ (w :: (a1, s)) ->\n\
           \                                      (y (<$\n\
           \                                            @ f\n\
           \                                            $dFunctor\n\
           \                                            @ a1\n\
           \                                            @ ()\n\
           \                                            (case w of { (a1, s') -> a1 })\n\
           \                                            z),\n\
           \                                       case w of { (a1, s') -> s' }) } in\n\
           \                              let {\n\
           \                                m1 :: StateT s (Sem r) a1\n\
           \                                m1 = f @ a1 @ m1 e } in\n\
           \                              (\\ (s1 :: s) ->\n\
           \                                 fmap\n\
           \                                   @ m\n\
           \                                   $dFunctor\n\
           \                                   @ (a1, s)\n\
           \                                   @ (x, s)\n\
           \                                   lvl\n\
           \                                   ((((m1 `cast` <Co:5>) s1) `cast` <Co:5>) @ m $dMonad k))\n\
           \                              `cast` <Co:5>\n\
           \                              };\n\
           \                            SS @ n1 co n2 ->\n\
           \                              case a1 `cast` <Co:49> of\n\
           \                              { Yo @ f1 @ m1 @ a1 $dFunctor3 e1 s2 nt f2 v1 ->\n\
           \                              (\\ (s' :: s) ->\n\
           \                                 k @ (x, s)\n\
           \                                   (Union\n\
           \                                      @ r\n\
           \                                      @ (Sem r)\n\
           \                                      @ (x, s)\n\
           \                                      @ n1\n\
           \                                      n2\n\
           \                                      (Yo\n\
           \                                         @ (IndexOf r n1)\n\
           \                                         @ (Sem r)\n\
           \                                         @ (x, s)\n\
           \                                         @ (Compose ((,) s) f1)\n\
           \                                         @ m1\n\
           \                                         @ a1\n\
           \                                         ($fFunctorCompose\n\
           \                                            @ ((,) s) @ f1 ($fFunctor(,) @ s) $dFunctor3)\n\
           \                                         e1\n\
           \                                         ((s', s2) `cast` <Co:10>)\n\
           \                                         (\\ (@ x) (x1 :: Compose ((,) s) f1 (m1 x)) ->\n\
           \                                            let {\n\
           \                                              p :: (s, Sem (e : r) (f1 x))\n\
           \                                              p = case x1 `cast` <Co:10> of { (x, y) ->\n\
           \                                                  (x, nt @ x y)\n\
           \                                                  } } in\n\
           \                                            let {\n\
           \                                              ds :: s\n\
           \                                              ds = case p of { (x, ds) -> x } } in\n\
           \                                            let {\n\
           \                                              ds :: Sem (e : r) (f1 x)\n\
           \                                              ds = case p of { (ds, y) -> y } } in\n\
           \                                            (\\ (@ (m :: * -> *))\n\
           \                                               ($dMonad :: Monad m)\n\
           \                                               (eta :: forall x. Union r (Sem r) x -> m x) ->\n\
           \                                               fmap\n\
           \                                                 @ m\n\
           \                                                 ($p1Applicative @ m ($p1Monad @ m $dMonad))\n\
           \                                                 @ (s, f1 x)\n\
           \                                                 @ (Compose ((,) s) f1 x)\n\
           \                                                 ((poly_f4 @ x @ f1 @ s) `cast` <Co:15>)\n\
           \                                                 (((mutual2 @ e @ s @ r @ (f1 x) f ds ds)\n\
           \                                                   `cast` <Co:6>)\n\
           \                                                    @ m $dMonad eta))\n\
           \                                            `cast` <Co:12>)\n\
           \                                         (\\ (x :: Compose ((,) s) f1 a1) ->\n\
           \                                            case x `cast` <Co:9> of { (x, y) -> (f2 y, x) })\n\
           \                                         (\\ (@ x) (eta :: Compose ((,) s) f1 x) ->\n\
           \                                            v1 @ x (snd @ s @ (f1 x) (eta `cast` <Co:9>))))))\n\
           \                              `cast` <Co:5>\n\
           \                              }\n\
           \                          }\n\
           \                          }))\n\
           \                    `cast` <Co:4>)\n\
           \                     s)\n\
           \        "#)
  ```
  </details>

  **The `solveSimpleWanteds` failure goes away if I use a version of polysemy-plugin that includes https://github.com/isovector/polysemy/pull/87** (version 0.2.0.2 doesn't have that change yet).

**Note that I'm using Nix**, but I made sure to test this with the exact same versions as the stack.yaml file gives, so LTS 13.0 and GHC 8.6.3. But still it seems that CI succeeded without problems, so Nix might give problems after all, but I sure as hell don't see how.

Even using LTS 13.24 and GHC 8.6.5 had the same errors though, so I don't think that's related.

Also note: I can build polysemy 0.2.0.0 with polysemy-plugin 0.2.0.1 without problems and the tests succeed, with the exact same dependencies as in the failures above. So something inbetween 0.2.0.0/plugin-0.2.0.1 and 0.3.0.0/plugin-0.2.0.2 broke the tests for me. I have yet to bisect this though.

--- a/polysemy-plugin/package.yaml
+++ b/polysemy-plugin/package.yaml
@@ -22,6 +22,7 @@ dependencies:
 - polysemy >= 0.1
 - syb >= 0.7 && <0.8
 - transformers >= 0.5.5.0 && < 0.6
+- containers >= 0.6 && <= 0.7
 
 library:
   source-dirs: src
@@ -41,7 +42,7 @@ tests:
     - polysemy-plugin
     - hspec >= 2.6.0 && < 3
     - should-not-typecheck >= 2.1.0 && < 3
-    - inspection-testing >= 0.4.1.1 && < 0.5
+    - inspection-testing >= 0.4.2 && < 0.5
 
 default-extensions:
   - DataKinds

--- a/polysemy-plugin/polysemy-plugin.cabal
+++ b/polysemy-plugin/polysemy-plugin.cabal
@@ -4,7 +4,7 @@ cabal-version: 1.12
 --
 -- see: https://github.com/sol/hpack
 --
--- hash: aceeac8482d66cadbcf9559c89ab7ff683f188490a14d0701c2a56b73fcd9b95
+-- hash: a515749ff182fca310bc1890e96ce0b182b3596261ba5db4661be3c44717f4c7
 
 name:           polysemy-plugin
 version:        0.2.0.2
@@ -40,6 +40,7 @@ library
   default-extensions: DataKinds DeriveFunctor FlexibleContexts GADTs LambdaCase PolyKinds RankNTypes ScopedTypeVariables StandaloneDeriving TypeApplications TypeOperators TypeFamilies UnicodeSyntax
   build-depends:
       base >=4.7 && <5
+    , containers >=0.6 && <=0.7
     , ghc >=8.6.3 && <8.7
     , ghc-tcplugins-extra >=0.3 && <0.4
     , polysemy >=0.1
@@ -64,10 +65,11 @@ test-suite polysemy-plugin-test
   ghc-options: -threaded -rtsopts -with-rtsopts=-N -fplugin=Polysemy.Plugin -O2
   build-depends:
       base >=4.7 && <5
+    , containers >=0.6 && <=0.7
     , ghc >=8.6.3 && <8.7
     , ghc-tcplugins-extra >=0.3 && <0.4
     , hspec >=2.6.0 && <3
-    , inspection-testing >=0.4.1.1 && <0.5
+    , inspection-testing >=0.4.2 && <0.5
     , polysemy >=0.3
     , polysemy-plugin
     , should-not-typecheck >=2.1.0 && <3

--- a/polysemy-plugin/src/Polysemy/Plugin/Fundep.hs
+++ b/polysemy-plugin/src/Polysemy/Plugin/Fundep.hs
@@ -36,6 +36,7 @@ import           Class
 import           CoAxiom
 import           Control.Monad
 import           Data.Bifunctor
+import           Data.Function (on)
 import           Data.IORef
 import           Data.List
 import           Data.Maybe
@@ -100,12 +101,21 @@ canUnify wanted given =
         ]
 
 
-mkWanted :: Bool -> CtLoc -> Type -> Type -> TcPluginM (Maybe Ct)
+mkWanted
+    :: Bool
+    -> CtLoc
+    -> Type
+    -> Type
+    -> TcPluginM (Maybe ( (OrdType, OrdType)  -- the types we want to unify
+                        , Ct                  -- the constraint
+                        ))
 mkWanted must_unify loc wanted given =
   if (not must_unify || canUnify wanted given)
      then do
        (ev, _) <- unsafeTcPluginTcM $ runTcSDeriveds $ newWantedEq loc Nominal wanted given
-       pure $ Just $ CNonCanonical ev
+       pure $ Just ( (OrdType wanted, OrdType given)
+                   , CNonCanonical ev
+                   )
      else
        pure Nothing
 
@@ -119,28 +129,23 @@ countLength eq as =
 
 
 ------------------------------------------------------------------------------
--- | 'Ct's don't have an 'Ord' or 'Hashable' or anything useful for comparing
--- for equality. So instead we put cheat by pretty-printing the constraint, and
--- using _that_ as our unique key. It's gross, but it works. Using 'FastString'
--- here gives us O(1) comparisons.
---
--- Use 'hashWanted' to get one of these.
-newtype HashedCt = HashedCt
-  { getHashedCt :: FastString
+-- | 'Type's don't have 'Eq' or 'Ord' instances by default, even though there
+-- are functions in GHC that implement these operations. This newtype gives us
+-- those instances.
+newtype OrdType = OrdType
+  { getOrdType :: Type
   }
-  deriving (Eq, Ord, Show, Outputable)
 
+instance Eq OrdType where
+  (==) = eqType `on` getOrdType
 
-hashWanted :: DynFlags -> Ct -> HashedCt
-hashWanted dyn ct =
-  HashedCt
-    . fsLit
-    . renderWithStyle dyn (ppr . ctev_pred $ cc_ev ct)
-    $ defaultDumpStyle dyn
+instance Ord OrdType where
+  compare = nonDetCmpType `on` getOrdType
+
 
 
 solveFundep
-    :: (IORef (S.Set HashedCt), Class)
+    :: (IORef (S.Set (OrdType, OrdType)), Class)
     -> [Ct]
     -> [Ct]
     -> [Ct]
@@ -166,12 +171,11 @@ solveFundep (ref, effCls) giv _ want = do
     already_emitted <- tcPluginIO $ readIORef ref
 
     let eqs' = catMaybes eqs
-        hashes = fmap (hashWanted dyn) eqs'
-        new = filter (not . flip S.member already_emitted . snd)
-            $ zip eqs' hashes
+        new = filter (not . flip S.member already_emitted . fst) eqs'
 
-    tcPluginIO $ modifyIORef ref $ S.union $ S.fromList $ fmap snd new
+
+    tcPluginIO $ modifyIORef ref $ S.union $ S.fromList $ fmap fst new
     -- pprTraceM "emitting" $ ppr new
 
     pure . TcPluginOk []
-         $ fmap fst new
+         $ fmap snd new

--- a/polysemy-plugin/src/Polysemy/Plugin/Fundep.hs
+++ b/polysemy-plugin/src/Polysemy/Plugin/Fundep.hs
@@ -41,12 +41,11 @@ import           Data.IORef
 import           Data.List
 import           Data.Maybe
 import qualified Data.Set as S
-import           FastString (FastString, fsLit)
-import           GHC (DynFlags, ModuleName)
+import           FastString (fsLit)
+import           GHC (ModuleName)
 import           GHC.TcPluginM.Extra (lookupModule, lookupName)
 import           Module (mkModuleName)
 import           OccName (mkTcOcc)
-import           Outputable
 import           TcPluginM (TcPluginM, tcLookupClass, tcPluginIO)
 import           TcRnTypes
 import           TcSMonad hiding (tcLookupClass)
@@ -167,15 +166,10 @@ solveFundep (ref, effCls) giv _ want = do
             _                 -> pure Nothing
         Just eff' -> mkWanted True loc eff eff'
 
-    dyn <- unsafeTcPluginTcM getDynFlags
     already_emitted <- tcPluginIO $ readIORef ref
+    let new_wanteds = filter (not . flip S.member already_emitted . fst)
+                    $ catMaybes eqs
 
-    let eqs' = catMaybes eqs
-        new = filter (not . flip S.member already_emitted . fst) eqs'
+    tcPluginIO $ modifyIORef ref $ S.union $ S.fromList $ fmap fst new_wanteds
+    pure . TcPluginOk [] $ fmap snd new_wanteds
 
-
-    tcPluginIO $ modifyIORef ref $ S.union $ S.fromList $ fmap fst new
-    -- pprTraceM "emitting" $ ppr new
-
-    pure . TcPluginOk []
-         $ fmap snd new

--- a/polysemy-plugin/src/Polysemy/Plugin/Fundep.hs
+++ b/polysemy-plugin/src/Polysemy/Plugin/Fundep.hs
@@ -1,3 +1,5 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
 ------------------------------------------------------------------------------
 -- The MIT License (MIT)
 --
@@ -30,22 +32,25 @@
 
 module Polysemy.Plugin.Fundep (fundepPlugin) where
 
-import Class
-import CoAxiom
-import Control.Monad
-import Data.Bifunctor
-import Data.List
-import Data.Maybe
-import FastString (fsLit)
-import GHC (ModuleName)
-import GHC.TcPluginM.Extra (lookupModule, lookupName)
-import Module (mkModuleName)
-import OccName (mkTcOcc)
-import TcPluginM (TcPluginM, tcLookupClass)
-import TcRnTypes
-import TcSMonad hiding (tcLookupClass)
-import TyCoRep (Type (..))
-import Type
+import           Class
+import           CoAxiom
+import           Control.Monad
+import           Data.Bifunctor
+import           Data.IORef
+import           Data.List
+import           Data.Maybe
+import qualified Data.Set as S
+import           FastString (FastString, fsLit)
+import           GHC (DynFlags, ModuleName)
+import           GHC.TcPluginM.Extra (lookupModule, lookupName)
+import           Module (mkModuleName)
+import           OccName (mkTcOcc)
+import           Outputable
+import           TcPluginM (TcPluginM, tcLookupClass, tcPluginIO)
+import           TcRnTypes
+import           TcSMonad hiding (tcLookupClass)
+import           TyCoRep (Type (..))
+import           Type
 
 
 polysemyInternalUnion :: ModuleName
@@ -56,7 +61,8 @@ fundepPlugin = TcPlugin
     { tcPluginInit = do
         md <- lookupModule polysemyInternalUnion (fsLit "polysemy")
         monadEffectTcNm <- lookupName md (mkTcOcc "Find")
-        tcLookupClass monadEffectTcNm
+        (,) <$> tcPluginIO (newIORef S.empty)
+            <*> tcLookupClass monadEffectTcNm
     , tcPluginSolve = solveFundep
     , tcPluginStop = const (return ()) }
 
@@ -111,9 +117,29 @@ countLength eq as =
   let grouped = groupBy eq as
    in zipWith (curry $ bimap head length) grouped grouped
 
-solveFundep :: Class -> [Ct] -> [Ct] -> [Ct] -> TcPluginM TcPluginResult
+
+newtype HashedCt = HashedCt
+  { getHashedCt :: FastString
+  }
+  deriving (Eq, Ord, Show, Outputable)
+
+
+hashWanted :: DynFlags -> Ct -> HashedCt
+hashWanted dyn ct =
+  HashedCt
+    . fsLit
+    . renderWithStyle dyn (ppr . ctev_pred $ cc_ev ct)
+    $ defaultDumpStyle dyn
+
+
+solveFundep
+    :: (IORef (S.Set HashedCt), Class)
+    -> [Ct]
+    -> [Ct]
+    -> [Ct]
+    -> TcPluginM TcPluginResult
 solveFundep _ _ _ [] = pure $ TcPluginOk [] []
-solveFundep effCls giv _ want = do
+solveFundep (ref, effCls) giv _ want = do
     let wantedEffs = allMonadEffectConstraints effCls want
         givenEffs = snd <$> allMonadEffectConstraints effCls giv
         num_wanteds_by_r = countLength eqType $ fmap (thd . snd) wantedEffs
@@ -129,5 +155,16 @@ solveFundep effCls giv _ want = do
             _                 -> pure Nothing
         Just eff' -> mkWanted True loc eff eff'
 
-    pure $ TcPluginOk [] $ catMaybes eqs
+    dyn <- unsafeTcPluginTcM getDynFlags
+    already_emitted <- tcPluginIO $ readIORef ref
 
+    let eqs' = catMaybes eqs
+        hashes = fmap (hashWanted dyn) eqs'
+        new = filter (not . flip S.member already_emitted . snd)
+            $ zip eqs' hashes
+
+    tcPluginIO $ modifyIORef ref $ S.union $ S.fromList $ fmap snd new
+    -- pprTraceM "emitting" $ ppr new
+
+    pure . TcPluginOk []
+         $ fmap fst new

--- a/polysemy-plugin/src/Polysemy/Plugin/Fundep.hs
+++ b/polysemy-plugin/src/Polysemy/Plugin/Fundep.hs
@@ -118,6 +118,13 @@ countLength eq as =
    in zipWith (curry $ bimap head length) grouped grouped
 
 
+------------------------------------------------------------------------------
+-- | 'Ct's don't have an 'Ord' or 'Hashable' or anything useful for comparing
+-- for equality. So instead we put cheat by pretty-printing the constraint, and
+-- using _that_ as our unique key. It's gross, but it works. Using 'FastString'
+-- here gives us O(1) comparisons.
+--
+-- Use 'hashWanted' to get one of these.
 newtype HashedCt = HashedCt
   { getHashedCt :: FastString
   }

--- a/polysemy-plugin/test/BadSpec.hs
+++ b/polysemy-plugin/test/BadSpec.hs
@@ -5,6 +5,7 @@
 module BadSpec where
 
 import Polysemy
+import Polysemy.State
 import Test.Hspec
 import Test.ShouldNotTypecheck
 
@@ -21,6 +22,9 @@ negativePos :: Member (KVStore String v) r => Sem r (Maybe Bool)
 negativePos = do
   getKV "hello"
 
+badState :: Member (State a) r => Sem r ()
+badState = put ()
+
 
 spec :: Spec
 spec = do
@@ -29,4 +33,6 @@ spec = do
       shouldNotTypecheck positivePos
     it "should not typecheck in negative position" $ do
       shouldNotTypecheck negativePos
+    it "should not typecheck badly polymorphic State" $ do
+      shouldNotTypecheck badState
 

--- a/polysemy-plugin/test/PluginSpec.hs
+++ b/polysemy-plugin/test/PluginSpec.hs
@@ -12,6 +12,7 @@ import Polysemy.Error
 import Polysemy.State
 import Polysemy.Output
 import Test.Hspec
+import Unsafe.Coerce
 
 
 
@@ -55,6 +56,15 @@ newtype MyString = MyString String
   deriving (IsString, Eq, Show)
 
 
+data Janky = forall s. Janky (forall i. Sem '[State s] ())
+
+jankyState :: Janky
+jankyState = Janky $ put True
+
+unsafeUnjank :: Janky -> Sem '[State Bool] ()
+unsafeUnjank (Janky sem) = unsafeCoerce sem
+
+
 spec :: Spec
 spec = do
   describe "State effect" $ do
@@ -95,6 +105,10 @@ spec = do
 
         it "should interpret against MyString" $ do
           flipShouldBe ("hello" :: MyString, ()) . run $ runState "nothing" oStrState
+
+    describe "existential state" $ do
+      it "JankyState should compile" $ do
+        flipShouldBe (True, ()) . run $ runState False $ unsafeUnjank jankyState
 
 
   describe "Error effect" $ do

